rfd/0163-vnet.md

```diff
@@ -0,0 +1,792 @@
+
+A green indicator means that a tunnel is currently open. The absence of an indicator means the
+tunnel was opened but has since been closed due to inactivity. An error indicator means that a
+connection for the given app could not be established. Clicking on the row with the error shows the
+full error message.
+
+<img width="871" alt="VNet tab" src="assets/0163-vnet-tab.png">
+
+#### Lifecycle
+
+The first version of VNet integration in Connect behaves similarly to Connect My Computer. After the
+user starts VNet and until they explicitly stop it, VNet starts automatically when launching the
+app. It prompts for a password for root privileges through osascript on each start.
+
+If VNet is running and the user logs in to another cluster, Connect prompts for credentials again
+and configures VNet for that cluster. If this attempt fails, VNet continues to work but the
+indicator on the connection list icon shows a red exclamation mark. Upon opening the VNet tab from
+the connection list, the user can learn what the error was and either retry the attempt or ignore
+the error.
+
+Unfortunately, [Touch ID prompts are reserved for Apple
+clients](https://forums.developer.apple.com/forums/thread/701118?answerId=705661022#705661022). We
+assume that typically the user is going to start VNet and then just leave it running, somewhat
+negating the downside of having to enter credentials on each start.
+
+##### Manual start and stop
+
+Starting VNet by clicking the item in the connection list in the top left starts VNet for all root
+clusters and turns on autostart for all of them. Stopping VNet from the connection list stops it for
+all root clusters and turns off autostart for all of them.
+
+If VNet is not running, starting VNet from the VNet tab starts it only for the root cluster of the
+current workspace. Likewise, stopping it from the VNet tab stops it only for the current root
+cluster and turns off autostart just for this cluster.
+
+This behavior aligns with a couple of ideas:
+
+1) The connection list is a UI element that is not scoped to any workspace, unlike tabs which are
+always scoped to a specific workspace.
+2) The user can turn off VNet for a specific workspace if they do not wish to use it in that
+workspace.
+
+##### Expired certs
+
+As shown in the diagram in the [DNS queries and IP address
+assignment](#dns-queries-and-ip-address-assignment) section, there are two calls made to the
+cluster, one to get the details of an app and one to proxy a connection to the cluster. If any of
+those calls fails due to expired certs, Connect prompts the user to log in again. It's similar to
+how Connect prompts for relogin after the cert of a local db proxy expires.
+
+##### Considered alternatives
+
+###### Launch VNet on system start
+
+Doing so would require some kind of persistent UI that is also launched on system start, perhaps an
+app in the Menu Bar. The UI would need to enable the user to manage the state of VNet and provide a
```

Comment:
Not sure if it would really require a persistent UI, whichever process launches in the background on start should be able to open Connect when it needs a UI

Reply:
Yeah, you had a good idea there with deep links. Though despite not needing a persistent UI, the VNet manager would have to be started from some kind of process that can later communicate with the Electron app.

Otherwise the Electron app wouldn't be able to tell VNet to stop/start, see recent connections and so on.

I'm going to reword this, but it still seems like a v2+ feature.